### PR TITLE
Added CSPDarkNet variants and as_backbone as a util

### LINKED ISSUE #1171: CSPDarkNet has weights only for when depth and width multiplier are 1.0
This makes the current weights compatible only with YoloX_l. 

Would it make sense to have CSPDarknet_tiny, _s, _m, _l and _x so that the imagenet weights can be trained for these separately?

cc. @ianstenbit @LukeWood

### PR DESCRIPTION
**Pull request by quantumalaviya**
# What does this PR do?
This PR fixes the issue discussed in #1171 and also migrates as_backbone to be a model util

Fixes #1171

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@LukeWood

### CODE REVIEW

**Comment by quantumalaviya** [comment]
@ianstenbit When you get a chance, can you review this and train weights for these variants after merge? To start with, weights for `CSPDarkNetTiny` would be enough for me to continue experiments on YoloX.

**Comment by quantumalaviya** [comment]
@qlzh727 (tagging because of the todo note) @tanzhenyu I have attempted to add `as_backbone` as a util here. Please take a look and see if this works.

File: keras_cv/models/utils_test.py
Comment (ianstenbit) [inline keras_cv/models/utils_test.py:32]: Could you rename this test to test_as_backbone_missing_backbone_level_outputs

File: keras_cv/models/csp_darknet.py
Comment (ianstenbit) [inline keras_cv/models/csp_darknet.py:375]: Should this be XL instead of X, or is X a convention for CSP Darket

**Review by ianstenbit — COMMENTED** [review]
Looks good -- just a few nits

**Comment by DavidLandup0** [comment]
Thanks for the PR!
In a previous PR, I was trying to use ResNetV2's `as_backbone()` but the issue was that I could not select the specific layers (only a range). The range had two issues:

- Unused layers (only needed the first and last, middle ones added params but weren't used)
- The final output had 2048 channels (while the paper used a few layers earlier to limit the channels to 256 to reduce the effect of the feature maps when added to another stage)

It's *great* that we can have `as_backbone()` as a general util instead of a method in each model. To make it more globally usable (and the default approach to exposing backbones) - should we maybe allow for `layer_names=('l1', 'l2', ...)` instead of a range?

File: keras_cv/models/csp_darknet.py
Comment (quantumalaviya) [reply keras_cv/models/csp_darknet.py:375]: The YoloX model corresponding to this is called YoloX-x. Hence, the name.

**Comment by quantumalaviya** [comment]
I agree with David. I think the API should be updated to something like this:

```python
as_backbone(levels=None)
```

Where `levels` is one of:
- a tuple of size 2 indictating a range which would work as it does currently. `None` at any one of the indices would imply the limit for that level. For example, (None, 5) would mean levels 3, 4, 5 for a backbone where 2 is the minimum level.
- a string indicating the name of the layer to output.
- a list of strings that have all the layer names required in the output.

Defaults to None which includes all levels from `_backbone_level_outputs`.

With this, all the models can potentially support `as_backbone`. `_backbone_level_outputs` will be required only for a model to support the range tuple. Moreover, this way, users don't necessarily have to know what the internal layer names are.

Ofc, this API would need some refining but maybe its a better alternative? Would love your opinions on this @LukeWood @ianstenbit @tanzhenyu

**Comment by ianstenbit** [comment]
> I agree with David. I think the API should be updated to something like this:
> 
> ```python
> as_backbone(levels=None)
> ```
> 
> Where `levels` is one of:
> 
> * a tuple of size 2 indictating a range which would work as it does currently. `None` at any one of the indices would imply the limit for that level. For example, (None, 5) would mean levels 3, 4, 5 for a backbone where 2 is the minimum level.
> * a string indicating the name of the layer to output.
> * a list of strings that have all the layer names required in the output.
> 
> Defaults to None which includes all levels from `_backbone_level_outputs`.
> 
> With this, all the models can potentially support `as_backbone`. `_backbone_level_outputs` will be required only for a model to support the range tuple. Moreover, this way, users don't necessarily have to know what the internal layer names are.
> 
> Ofc, this API would need some refining but maybe its a better alternative? I also wonder if this util is should even be exposed or kept as an internal util. Would love your opinions on this @LukeWood @ianstenbit @tanzhenyu

I like the idea of adding some more flexibility to the `as_backbone` API -- can you open up an issue to discuss this?

**Review by ianstenbit — APPROVED** [review]
LGTM -- thanks!

As long as the GPU tests pass, I'll merge this and kick off training of the CSPDarkNetTiny to unblock you on YoloX

**Comment by tanzhenyu** [comment]
> I agree with David. I think the API should be updated to something like this:
> 
> ```python
> as_backbone(levels=None)
> ```
> 
> Where `levels` is one of:
> 
> * a tuple of size 2 indictating a range which would work as it does currently. `None` at any one of the indices would imply the limit for that level. For example, (None, 5) would mean levels 3, 4, 5 for a backbone where 2 is the minimum level.
> * a string indicating the name of the layer to output.
> * a list of strings that have all the layer names required in the output.
> 
> Defaults to None which includes all levels from `_backbone_level_outputs`.
> 
> With this, all the models can potentially support `as_backbone`. `_backbone_level_outputs` will be required only for a model to support the range tuple. Moreover, this way, users don't necessarily have to know what the internal layer names are.
> 
> Ofc, this API would need some refining but maybe its a better alternative? I also wonder if this util is should even be exposed or kept as an internal util. Would love your opinions on this @LukeWood @ianstenbit @tanzhenyu

What we returned here is dict: int -> output_tensor. If we do support string names, I have two questions:
1) how do expect users to find those layer names -- this can sometimes be trivial, but sometimes quite some manual work
2) would the output be string -> output_tensor, and how does that interact with our FPN / Detection / Segmentation models which today only accept int?

**Pull request merged**
